sdk/core/Azure.Core/src/Shared/BearerTokenChallengeAuthenticationPolicy.cs

```diff
@@ -0,0 +1,427 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Diagnostics;
+
+#nullable enable
+
+namespace Azure.Core.Pipeline
+{
+    /// <summary>
+    /// A policy that sends an <see cref="AccessToken"/> provided by a <see cref="TokenCredential"/> as an Authentication header.
+    /// </summary>
+    internal class BearerTokenChallengeAuthenticationPolicy : HttpPipelinePolicy
+    {
+        private const string AuthenticationChallengePattern = @"(\w+) ((?:\w+="".*?""(?:, )?)+)(?:, )?";
+        private const string ChallengeParameterPattern = @"(?:(\w+)=""([^""]*)"")+";
+        private const string ChallengeHeader = "WWW-Authenticate";
+        private static readonly Regex s_AuthenticationChallengeRegex = new Regex(AuthenticationChallengePattern, RegexOptions.Compiled);
+        private static readonly Regex s_ChallengeParameterRegex = new Regex(ChallengeParameterPattern, RegexOptions.Compiled);
+
+        private readonly AccessTokenCache _accessTokenCache;
+        private string[] _scopes;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="BearerTokenChallengeAuthenticationPolicy"/> using provided token credential and scope to authenticate for.
+        /// </summary>
+        /// <param name="credential">The token credential to use for authentication.</param>
+        /// <param name="scope">The scope to authenticate for.</param>
+        public BearerTokenChallengeAuthenticationPolicy(TokenCredential credential, string scope) : this(credential, new[] { scope }) { }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="BearerTokenChallengeAuthenticationPolicy"/> using provided token credential and scopes to authenticate for.
+        /// </summary>
+        /// <param name="credential">The token credential to use for authentication.</param>
+        /// <param name="scopes">Scopes to authenticate for.</param>
+        public BearerTokenChallengeAuthenticationPolicy(TokenCredential credential, IEnumerable<string> scopes)
+            : this(credential, scopes, TimeSpan.FromMinutes(5), TimeSpan.FromSeconds(30)) { }
```

Comment:
These times aren't configurable, at least by a service? Or just no reason yet?

Reply:
We could expose them as configurable, I suppose. This is taken directly from BearerTokenAuthenticationPolicy

https://github.com/Azure/azure-sdk-for-net/blob/d9829c9d298abbbcb51f4a5db8178591d7ea2e8c/sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs#L32-L33

Reply:
These parameters `tokenRefreshOffset` and `tokenRefreshRetryDelay` configure how the far in advance we will refresh a cached token, and how often we'll retry to refresh the token in the case the refresh fails or doesn't give us a new token. I could imagine end users possibly wanting to configure them based on their environment. For instance, a client with a flakey connection to AAD might increase the refresh offset to more pro-actively refresh tokens. But I don't think client libraries would have any way to meaningfully configure them as they are not service specific.  Even the scenarios for end-users configuring these are tenuous at best, because I don't know that we could give people meaningful guidance on how these could be configured to improve reliability in situations such as the one I mentioned. So as you surmised we haven't made them externally configurable because we haven't needed to yet. The fact that their first class constructor arguments on the internal  constructor called here is an artifact of making it configurable for testing.

Reply:
That makes sense. We can open them up later. Perhaps make them part of service-specific ClientOptions and have them flow-through, since this policy is an internal class added automatically by services that need it (like Key Vault or ARM).